src/System.Text.Json/src/System/Text/Json/Serialization/Converters/DefaultImmutableConverter.cs

```diff
@@ -0,0 +1,122 @@
+        }
+
+        public override IEnumerable CreateFromList(Type enumerableType, Type elementType, IList sourceList)
+        {
+            Debug.Assert(enumerableType.IsGenericType);
+
+            Type underlyingType = enumerableType.GetGenericTypeDefinition();
+            Type constructingType;
+            string name = underlyingType.FullName;
+
+            switch (name)
+            {
+                case ImmutableListGenericTypeName:
+                case ImmutableListGenericInterfaceTypeName:
+                    constructingType = typeof(ImmutableList);
+                    break;
+                case ImmutableArrayGenericTypeName:
+                    constructingType = typeof(ImmutableArray);
+                    break;
+                case ImmutableStackGenericTypeName:
+                case ImmutableStackGenericInterfaceTypeName:
+                    constructingType = typeof(ImmutableStack);
+                    break;
+                case ImmutableQueueGenericTypeName:
+                case ImmutableQueueGenericInterfaceTypeName:
+                    constructingType = typeof(ImmutableQueue);
+                    break;
+                case ImmutableSortedSetGenericTypeName:
+                    constructingType = typeof(ImmutableSortedSet);
+                    break;
+                case ImmutableHashSetGenericTypeName:
+                case ImmutableSetGenericInterfaceTypeName:
+                    constructingType = typeof(ImmutableHashSet);
+                    break;
+                default:
+                    return sourceList;
+            }
+
+            MethodInfo[] constructingTypeMethods = constructingType.GetMethods();
+            MethodInfo constructingMethod = null;
+
+            foreach (MethodInfo method in constructingTypeMethods)
+            {
+                if (method.Name == "CreateRange" && method.GetParameters().Length == 1)
+                {
+                    constructingMethod = method;
+                    break;
+                }
+            }
+
+            if (constructingMethod != null)
+            {
+                if (elementType == null)
+                {
+                    Type[] args = enumerableType.GetGenericArguments();
+                    Debug.Assert(args.Length == 1);
+
+                    elementType = args[0];
+                }
+
+                MethodInfo mi = typeof(DefaultImmutableConverter).GetMethod("CreateFromListInternal", BindingFlags.NonPublic | BindingFlags.Static);
+
+                if (mi != null)
+                {
+                    MethodInfo createImmutableMethod = constructingMethod.MakeGenericMethod(elementType);
+
+                    MethodInfo createFromListInternalMethod = mi.MakeGenericMethod(elementType);
+                    return (IEnumerable)createFromListInternalMethod.Invoke(null, new object[] { createImmutableMethod, sourceList });
+                }
+            }
+
+            return sourceList;
```

Comment:
So it looks like all of the possible error cases fall through to here. Is the idea that the serializer would throw for this case? Does it make sense to throw and exception here? It might even simplify the code a bit to handle these cases explicitly.